subprojects/core/src/main/java/org/gradle/execution/plan/PersistentList.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.execution.plan;
+
+import org.gradle.internal.Cast;
+
+import javax.annotation.CheckReturnValue;
+import java.util.function.Consumer;
+
+public interface PersistentList<T> {
```

Comment:
I'm not sure `Persistent` is a good name here. This is more like an `ImmutableLinkedList`, no?